File: python/abc363/c.py
from more_itertools import distinct_permutations
from numba import jit

num, kaibun_length = map(int, input().split())
base_string = input()

count = 0

@jit(nopython=True, cache=True)
def is_palindrome(s):
    for i in range(len(s) // 2):
        if s[i] != s[-i - 1]:
            return False
    return True

@jit(nopython=True)
def is_k_palindrome(s, k):
    for i in range(num - k + 1):
        if is_palindrome(s[i:i + k]):
            return True
    return False

count = sum(
    not is_k_palindrome(perm, kaibun_length) for perm in distinct_permutations(base_string)
)

print(count)

File: cpp/abc363/c.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <unordered_set>
#include <vector>

using namespace std;

// 回文かどうかをチェックする関数
bool isPalindrome(const string& s) {
    int left = 0;
    int right = s.size() - 1;
    while (left < right) {
        if (s[left] != s[right]) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

// k文字の回文を含むかどうかをチェックする関数
bool containsKLengthPalindrome(const string& s, int k) {
    for (int i = 0; i <= s.size() - k; ++i) {
        if (isPalindrome(s.substr(i, k))) {
            return true;
        }
    }
    return false;
}

int main() {
    int num, kaibun_length;
    cin >> num >> kaibun_length;
    string base_string;
    cin >> base_string;

    int count = 0;

    sort(base_string.begin(), base_string.end());

    do {
        if (!containsKLengthPalindrome(base_string, kaibun_length)) {
            count++;
        }
    } while (next_permutation(base_string.begin(), base_string.end()));

    cout << count << endl;

    return 0;
}
